R/write_output.R

```diff
@@ -0,0 +1,336 @@
+        output_dir,
+        job_id,
+        "tasks",
+        task_id,
+        "metadata.json"
+      )
+      jsonlite::write_json(metadata, metadata_path, pretty = TRUE)
+      cli::cli_alert_success("Wrote metadata to {.path {metadata_path}}")
+    },
+    error = function(cnd) {
+      # Downgrade erroring out to a warning so we can catch and log
+      cli::cli_warn(
+        "Failure writing outputs",
+        parent = cnd,
+        class = "no_outputs"
+      )
+    }
+  )
+
+  invisible(NULL)
+}
+
+#' Create output directory structure for a given job and task.
+#'
+#' This function generates the necessary directory structure for storing output
+#' files related to a job and its tasks, including directories for raw samples
+#' and summarized quantiles.
+#'
+#' @param output_dir String. The base output directory path.
+#' @param job_id String. The identifier for the job.
+#' @param task_id String. The identifier for the task.
+#'
+#' @return The path to the base output directory (invisible).
+#' @export
+write_output_dir_structure <- function(output_dir, job_id, task_id) {
+  # Top level
+  dir.create(output_dir,
+    showWarnings = FALSE
+  )
+  # Second level
+  dir.create(file.path(output_dir, job_id),
+    showWarnings = FALSE
+  )
+  # Third level
+  dir.create(file.path(output_dir, job_id, "tasks"),
+    showWarnings = FALSE
+  )
+  dir.create(file.path(output_dir, job_id, "samples"),
+    showWarnings = FALSE
+  )
+  dir.create(file.path(output_dir, job_id, "summaries"),
+    showWarnings = FALSE
+  )
+
+  # Fourth level
+  dir.create(file.path(output_dir, job_id, "tasks", task_id),
+    showWarnings = FALSE
+  )
+
+  invisible(output_dir)
+}
+
+#' Extract posterior draws from a Stan fit object.
+#'
+#' This function extracts posterior draws for specific parameters from a Stan
+#' fit object and prepares a fact table containing unique date-time-parameter
+#' combinations for further merging.
+#'
+#' @param fit A Stan fit object with posterior estimates.
+#'
+#' @return A list containing two elements: `stan_draws` (the extracted draws in
+#' long format) and `fact_table` (a table of unique date-time-parameter
+#' combinations).
+#' @noRd
+extract_draws_from_fit <- function(fit) {
+  # Step 1: Extract unique date-time-parameter combinations
+  fact_table <- fit[["estimates"]][["samples"]][,
+    c("date", "time", "parameter"),
+    with = FALSE
+  ]
+  fact_table <- stats::na.omit(unique(fact_table))
+
+  # Step 1.1: Add corresponding 'obs_cases' rows for 'latent_cases' dates
+  obs_fact_table <- fact_table[
+    fact_table[["parameter"]] == "imputed_reports",
+  ]
+  # The EpiNow2 summary table has the variable `imputed_reports`
+  # for nowcast-corrected cases, but not `obs_reports` for right-
+  # truncated cases to compare to the observed data. We want both.
+  #
+  # The dates for `obs_reports` are the same as for `imputed_reports`
+  # (their differences are the nowcast correction + error structure).
```

Comment:
i.e this is correct if you interpret it as imputed reports have error and are corrected for right truncation (i.e its a posterior prediction of the final observed value) whilst obs_reports is the expected value actually observed in real time (but without obs error)

Reply:
https://github.com/CDCgov/cfa-epinow2-pipeline/pull/41/commits/1c69a0a9e8341c1c0e171a0714be2f053bcb9f11